### PR TITLE
Begin type annotation

### PR DESCRIPTION
**Pull request by ConnorRigby**
We should probably remember to change noImplecitAny to false at some point.

### CODE REVIEW

File: src/config/actions.tsx [hunk 1]
```diff
@@ -12,7 +12,7 @@ export function changeApiPort(port: string): ReduxAction<ChangeApiPort> {
 };
 
 export function ready(e: Event) {
-  return (dispatch, getState) => {
+  return (dispatch: Function, getState: Function) => {
```
Comment (RickCarlino) [inline src/config/actions.tsx]: This is a `Thunk`.

File: src/config/reducer.tsx [hunk 2]
```diff
@@ -6,7 +6,7 @@ let stripUrl = (url: string) => url.replace(/(https|http|[0-9]|\:|\/|\/\/)/g, ""
 
 let initialState: ConfigState = {
   host: stripUrl(location.host),
-  port: location.host["includes"]("localhost") ? "3000" : (location.port || "80")
+  port: (location.host as any)["includes"] ("localhost") ? "3000" : (location.port || "80")
```
Comment (RickCarlino) [inline src/config/reducer.tsx:9]: This is actually a `string`.

File: src/controls/controls.tsx [hunk 3]
```diff
@@ -24,11 +24,11 @@ interface AxisInputBoxProps {
 export class AxisInputBox extends React.Component<AxisInputBoxProps, {}> {
 
   primary(): string {
-    return this.props.bot.axisBuffer[this.props.axis];
+    return this.props.bot.axisBuffer[this.props.axis] || "";
   }
 
   secondary(): string {
-    let num = this.props.bot.hardware[this.props.axis];
+    let num = (this.props.bot.hardware as any)[this.props.axis];
```
Comment (RickCarlino) [inline src/controls/controls.tsx:31]: Could this be a `Map<string>` or `{[name: string]: string}`?

Comment (RickCarlino) [reply src/controls/controls.tsx:31]: Not sure 100%, actually.

File: src/controls/direction_button.tsx [hunk 4]
```diff
@@ -7,7 +7,7 @@ export class DirectionButton extends React.Component<any, any> {
     let isNegative = (this.props.direction === "up") ||
                       (this.props.direction === "right");
     let multiplier = (isNegative) ? -1 : 1;
-    payload[this.props.axis] = (this.props.steps || 250 ) * multiplier;
+    (payload as any)[this.props.axis] = (this.props.steps || 250 ) * multiplier;
```
Comment (RickCarlino) [inline src/controls/direction_button.tsx:10]: Could be `number`, right?

File: src/controls/toggle_button.tsx [hunk 5]
```diff
@@ -2,17 +2,21 @@ import * as React from "react";
...
       "true":      "yes",
       "on":        "yes",
-      "undefined": "---"
+      "undefined": "---",
     };
-
-    return captions[String(this.props.toggleval)] || "---";
+    let togval = String(this.props.toggleval);
+    if(togval){
+      return (captions as any) [togval];
```
Comment (RickCarlino) [inline src/controls/toggle_button.tsx:16]: Probably `as string`.

File: src/controls/toggle_button.tsx [hunk 6]
```diff
@@ -30,7 +34,7 @@ export class ToggleButton extends React.Component<any, any> {
       "undefined": yellowCSS
     };
 
-    return cssClasses[String(this.props.toggleval)] || yellowCSS;
+    return (cssClasses as any)[String(this.props.toggleval)] || yellowCSS;
```
Comment (RickCarlino) [inline src/controls/toggle_button.tsx:37]: Same as above. `as string`.

File: src/generate_reducer.ts [hunk 7]
```diff
@@ -1,6 +1,6 @@
 import { ReduxAction } from "./interfaces";
 
-const NOOP = (s, a) => s;
+const NOOP = (s: any, a: ReduxAction<{}>) => s;
```
Comment (RickCarlino) [inline src/generate_reducer.ts:3]: I think this one is `(s: T, . . . . )`

Comment (RickCarlino) [reply src/generate_reducer.ts:3]: Actually, move `NOOP` into the function definition after the current first line.

File: src/sequences/actions.ts [hunk 8]
```diff
@@ -179,7 +179,7 @@ export function deleteSequence(index: number) {
   // misc errors 
   // dependency error. 
 
-  return function(dispatch, getState){
+  return function(dispatch: Function, getState: Function){
```
Comment (RickCarlino) [inline src/sequences/actions.ts:182]: This is a `Thunk`

File: tsconfig.json [hunk 9]
```diff
@@ -4,7 +4,7 @@
     "module": "commonjs",
     "jsx": "react",
     "experimentalDecorators": true,
-    "noImplicitAny": false,
+    "noImplicitAny": true,
```
Comment (RickCarlino) [inline tsconfig.json]: Once you implement all the suggestions, you can flip this back to normal.

**Review by RickCarlino — CHANGES_REQUESTED** [review]
Cool. Just a few things.

File: src/devices/interfaces.ts [hunk 10]
```diff
@@ -109,3 +109,9 @@ export interface HardwareState {
   pin13?: string;
 }
 
+export interface MqttMessage {
+  error? : string | undefined;
+  id?: string | undefined;
+  result: HardwareState;
+}
```
Comment (RickCarlino) [inline src/devices/interfaces.ts:116]: TODO: Delete this.

File: src/devices/message_catcher.ts [hunk 11]
```diff
@@ -66,7 +67,7 @@ export let catchMessage = (resp: any) => (choices: ResponseOutcome) => {
       error: isBotErrorResponse,
       notification: isNotification
     }).mapValues((f: Function) => f(resp));
-  let matchCount = (results.invert(true).value())["true"] || [];
+  let matchCount = (results.invert(true).value() as any)["true"] || [];
```
Comment (ConnorRigby) [inline src/devices/message_catcher.ts:70]: `let matchCount = (results.invert(true).value() as {[name: string}: any)["true"] || [];`

File: src/devices/reducer.tsx [hunk 12]
```diff
@@ -49,15 +49,15 @@ export let botReducer = generateReducer<BotState>(initialState)
     return state;
   })
   .add<any>("CHANGE_AXIS_BUFFER", function(state, action) {
-    let axisBuffer = _.assign({}, state.axisBuffer);
+    let axisBuffer: any = _.assign({}, state.axisBuffer);
```
Comment (RickCarlino) [inline src/devices/reducer.tsx]: This is defensive copying. No longer needs to be done.

File: src/devices/reducer.tsx [hunk 13]
```diff
@@ -49,15 +49,15 @@ export let botReducer = generateReducer<BotState>(initialState)
...
     axisBuffer[action.payload.key] = action.payload.val;
 
     return _.assign<any, BotState>({}, state, {
       axisBuffer: axisBuffer
     });
   })
   .add<any>("CHANGE_SETTINGS_BUFFER", function(state, action) {
-    let settingsBuffer = _.assign({}, state.settingsBuffer);
+    let settingsBuffer: any = _.assign({}, state.settingsBuffer);
```
Comment (RickCarlino) [inline src/devices/reducer.tsx]: Same thing. Looks like more defensive copying.

File: src/regimens/bulk_scheduler/reducer.ts [hunk 14]
```diff
@@ -41,54 +29,33 @@ function newState(index: number): BulkSchedulerState {
...
+  })
+  .add<any>("TOGGLE_DAY", function(state, action) {
+      let week = state.form.weeks[action.payload.week];
+      let day = `day${action.payload.day}`;
+      let days = (week.days as {[day: string]: boolean})
+      days[day] = !days[day];
+      return state;
+  })
+  .add<any>("COMMIT_BULK_EDITOR", function(state, action) {
+    return newState(state.currentRegimen);
```
Comment (RickCarlino) [inline src/regimens/bulk_scheduler/reducer.ts:56]: Is this just defensive copying? It might not be, but please check to make sure.

**Review by RickCarlino — COMMENTED** [review]
:clap: Looks good. See my comment and then let's merge this thang.

File: src/config/reducer.tsx (hunk 2 above)
Comment (RickCarlino) [reply src/config/reducer.tsx:9]: Not relevant. Nvm

**Pull request merged**